clang/lib/Sema/SemaStmtAttr.cpp

```diff
@@ -278,9 +278,11 @@ static IVDepExprResult HandleFPGAIVDepAttrExpr(Sema &S, Expr *E,
       return IVDepExprResult::Invalid;
     SafelenValue = ArgVal->getZExtValue();
     // ivdep attribute allows both safelen = 0 and safelen = 1 with a warning.
-    if (SafelenValue == 0 || SafelenValue == 1)
+    if (SafelenValue == 0 || SafelenValue == 1) {
       S.Diag(E->getExprLoc(), diag::warn_ivdep_attribute_argument)
           << SafelenValue;
+      return IVDepExprResult::Invalid;
```

Comment:
```suggestion
      return IVDepExprResult::Null;
```
Since it is not really invalid, but rather has no effect, Null may be better here.   Invalid could also trigger further diagnostics.

Reply:
@premanandrao, NULL wouldn't work here. NULL would give assertion like this:
```
llvm/include/llvm/Support/Casting.h:578: decltype(auto) llvm::cast(From*) [with To = clang::MemberExpr; From = const clang::Expr]: Assertion `isa<To>(Val) && "cast<Ty>() argument of incompatible type!"' failed.
````

We have this in our ivdep source codes:
```
static IVDepExprResult HandleFPGAIVDepAttrExpr(Sema &S, Expr *E,
                                               unsigned &SafelenValue) {
  if (!E)
    return IVDepExprResult::Null;

  if (E->isInstantiationDependent())
    return IVDepExprResult::Dependent;

  Optional<llvm::APSInt> ArgVal = E->getIntegerConstantExpr(S.getASTContext());
  if (ArgVal) {
    if (checkSYCLIntelFPGAIVDepSafeLen(S, *ArgVal, E))
      return IVDepExprResult::Invalid;
    SafelenValue = ArgVal->getZExtValue();
    // ivdep attribute allows both safelen = 0 and safelen = 1 with a warning.
    if (SafelenValue == 0 || SafelenValue == 1) {
      S.Diag(E->getExprLoc(), diag::warn_ivdep_attribute_argument)
          << SafelenValue;
      return IVDepExprResult::Invalid;
    }
    return IVDepExprResult::SafeLen;
  }

  if (isa<DeclRefExpr>(E) || isa<MemberExpr>(E)) {
    if (!E->getType()->isArrayType() && !E->getType()->isPointerType()) {
      S.Diag(E->getExprLoc(), diag::err_ivdep_declrefexpr_arg);
      return IVDepExprResult::Invalid;
    }
    return IVDepExprResult::Array;
  }

  S.Diag(E->getExprLoc(), diag::err_ivdep_unknown_arg);
  return IVDepExprResult::Invalid;
}
```

Reply:
> @premanandrao, We are returning NULL for no Ivdep expression value.

Isn't a value of 0 or 1 the same as no Ivdep expression value? i.e. don't these three mean the same:
[[intel::ivdep(1)]], [[intel::ivdep(0)]] and [[intel::ivdep]]

We seem to be returning NULL for the last case.  Why not for the first two also?

Reply:
I am sorry, you are right.  An empty ivdep means something else.

Reply:
>>An empty ivdep means something else.

Yes.

Thanks @premanandrao